File: lib/specter/codec_capability.ex
defmodule Specter.RtpCodecCapability do
  @moduledoc """
  A representation of webrtc.rs RTCRtpCodecCapability.
  """

  @typedoc """
  For the meaning of specific fields refer to
  https://w3c.github.io/webrtc-pc/#rtcrtpcodeccapability

  Additionaly, webrtc.rs allows to specify extra RTCP packet types.
  However, this is not implemented yet (and also not included in w3c standard).
  """
  @type t() :: %__MODULE__{
          mime_type: String.t(),
          clock_rate: non_neg_integer(),
          channels: non_neg_integer(),
          sdp_fmtp_line: String.t()
        }

  defstruct mime_type: "",
            clock_rate: 0,
            channels: 0,
            sdp_fmtp_line: ""
end

File: native/specter_nif/src/codec_capability.rs
use rustler::NifStruct;
use webrtc::rtp_transceiver::rtp_codec::RTCRtpCodecCapability;

#[derive(NifStruct)]
#[module = "Specter.RtpCodecCapability"]
pub struct RtpCodecCapability {
    pub mime_type: String,
    pub clock_rate: u32,
    pub channels: u16,
    pub sdp_fmtp_line: String,
}

impl From<&RtpCodecCapability> for RTCRtpCodecCapability {
    fn from(rtp_codec_capability: &RtpCodecCapability) -> Self {
        RTCRtpCodecCapability {
            mime_type: rtp_codec_capability.mime_type.clone(),
            clock_rate: rtp_codec_capability.clock_rate,
            channels: rtp_codec_capability.channels,
            sdp_fmtp_line: rtp_codec_capability.sdp_fmtp_line.clone(),
            ..Default::default()
        }
    }
}

impl From<RtpCodecCapability> for RTCRtpCodecCapability {
    fn from(rtp_codec_capability: RtpCodecCapability) -> Self {
        RTCRtpCodecCapability {
            mime_type: rtp_codec_capability.mime_type,
            clock_rate: rtp_codec_capability.clock_rate,
            channels: rtp_codec_capability.channels,
            sdp_fmtp_line: rtp_codec_capability.sdp_fmtp_line,
            ..Default::default()
        }
    }
}
